DIPs/DIP1xxx.md

```diff
@@ -0,0 +1,249 @@
+    static foreach (n; needles)
+        static assert(is(typeof(startsWith!pred(haystack, n))), 
+                "predicate `" ~ pred.stringof "` must be valid for `startsWith!pred(haystack, "~n.stringof~"`));
+}
+```
+
+### Reference
+
+https://github.com/dlang/phobos/pull/6607
+
+https://issues.dlang.org/show_bug.cgi?id=13683
+
+## Contents
+* [Rationale](#rationale)
+* [Description](#description)
+* [Breaking Changes and Deprecations](#breaking-changes-and-deprecations)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+
+It is well known that compilation error messages due to template contraint overload resolution 
+are particularly difficult to decipher. This is not helped by the number of overloads and very 
+precice (and thus complex) constraints placed on the overloads. When overload resolution fails
+the compiler will print out all the in scope overloads and their constraints, without indication
+of which constraints have failed.
+
+While it is not possible in the general case to provide useful information as to what constraints
+have failed and why, because a constraint may have an arbitrary combination of logic, the vast 
+majority of constraints are expressed in Conjunctive Normal Form (CNF). In this case it is definitely 
+possible to provide better daignostics as to which clauses have failed. However the current grammer
+provides no way to translate particularly verbose constraints to a user not intimately familiar with 
+the constraint.
+
+This DIP therefore proposes to formalise the use of CNF constraints by allowing multiple `if` constraints,
+the expression form with an optional message (similar to what was done with contracts in DIP1009), as well as block statements that 
+allows the use of `static foreach` and to declare `alias`es and `enum`s to eliminate the need for recursive templates in template constrains 
+(similar to the `in` contract form prior to DIP1009).
+This will put the compiler in a much better position to provide useful diagnostics, such as indicating which clauses are not satisfied 
+and allowing the template author to provide messages in the case of non-intuitive formulations of constraints
+e.g. `if(isForwardRange!(R) == isInputRange!(R), "needles that are ranges must be forward ranges")`.
+
+Using the particularly egregious example of the first overload of `std.algorithm.searching.countUntil`,
+its current signature of
+
+```D
+ptrdiff_t countUntil(alias pred = "a == b", R, Rs...)(R haystack, Rs needles)
+if (isForwardRange!R
+&& Rs.length > 0
+&& isForwardRange!(Rs[0]) == isInputRange!(Rs[0])
+&& is(typeof(startsWith!pred(haystack, needles[0])))
+&& (Rs.length == 1
+|| is(typeof(countUntil!pred(haystack, needles[1 .. $])))))
+```
+
+would be be written using the block statement form to elimiate the recursive constraint as  
+
+```D
+ptrdiff_t countUntil(alias pred = "a == b", R, Rs...)(R haystack, Rs needles)
+if(isForwardRange!R)
```

Comment:
the document would be more readable by folling phobos style. There numerous places where you didn't put a space between `if` and the opening paren.